docs/Software Engineering Blog/posts/Comparing_Human_and_AI_way_for_Big-String-Set_Intersection_Performance_in_Java_by_JMH.md

```diff
@@ -0,0 +1,115 @@
+
+### ChatGPT way
+
+The second competitor is ChatGPT (GPT4o).
+
+![chatGPT.png](resources/BigSetIntersec/chatGPT.png)
+
+ChatGpt write a straightforward function that using native java method `Set.retainAll` to do the trick. we call it the `smallSetRetainAllBigSet`.
+
+---
+
+After the AI handout their solution, let try some old-school way to find some answer. Here I found the [question](https://stackoverflow.com/questions/2851938/efficiently-finding-the-intersection-of-a-variable-number-of-sets-of-strings) had been asked 3 years before, In this question, the accepted answer is also using the `Set.retainAll`. but we still get two more way from other answer. Let's take a look.
+
+### Stackoverflow way1: stream.filter
+
+The first human provide answer is to use the Java 8 feature to open a stream() on the small set, and check if each element is contained by the big set 
+
+![stream.png](resources/BigSetIntersec/stream.png)
+
+So we called this the `streamSmallSetIsContainedByBigSet` and another version as `streamSmallSetIsContainedByNewBigSet` to compare with the Gemini way.
+
+### Stackoverflow way2: guava
+
+The final one way is to use the Google provide utils library, [Guava](https://github.com/google/guava). There is a method `intersection()` in the Utils-Class `Sets`. 
+
+![guava.png](resources/BigSetIntersec/guava.png)
+
+## How to test: JMH
+
+We all know that the JVM needs some time (code execution) to warm up for steady performance on methods. Besides, we also need to do performance calculations and produce easy-to-read results. It sounds very troublesome. But luckily, we have Java Microbenchmark Harness (JMH) to do that for us.
+
+### how to use
+
+There are lots of guidance or How-To articles online to do that. I follow the README in [jmh-gradle-plugin](https://github.com/melix/jmh-gradle-plugin) and every thing go well. If you are interested in the detail or code implementation, please refer to [my repository](https://github.com/NoahHsu/performance-benchmark/tree/main/src/jmh/java/org/example/bigset/intersacation), there is a working example.
+
+### scenario description
+
+In my use case, I need to extract user IDs that are present in both of two sets: one representing users who have previously purchased my product, and the other representing users who have agreed to receive advertisements.
+
+To evaluate performance, I designed two scenarios:
+
+1. Both sets contain 100,000 UUIDs, with 1,000 IDs as the intersection.
+2. One set contains 100,000 UUIDs, while the other contains 1,100 UUIDs, including the same 1,000 IDs as the intersection. 
+
+## Comparing Result
+
+Here, we can start to execute the JMH command by the [jmh-gradle-plugin](https://github.com/melix/jmh-gradle-plugin). However, there are some essential points that we should be cautious about.
```

Comment:
⚠️ **[vale]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[alex.ProfanityUnlikely] Be careful with 'execute', it's profane in some cases.